crates/amaru-ledger/src/rules/transaction/scripts.rs

```diff
@@ -0,0 +1,145 @@
+            .lookup(input)
+            .and_then(|output| match output {
+                amaru_kernel::PseudoTransactionOutput::PostAlonzo(transaction_output) => {
+                    transaction_output
+                        .script_ref
+                        .as_deref()
+                        .and_then(|script_ref| {
+                            // If there is a provided ScriptRef, make sure it is required by an input
+                            let hash = script_ref.script_hash();
+                            if required_scripts.contains(&hash) {
+                                Some(hash)
+                            } else {
+                                None
+                            }
+                        })
+                }
+                amaru_kernel::PseudoTransactionOutput::Legacy(_) => None,
+            })
+    })
+    .collect::<BTreeSet<_>>();
+
+    let provided_scripts: BTreeSet<_> = get_provided_scripts(witness_set)
+        .into_iter()
+        .chain(provided_reference_scripts)
+        .collect();
+
+    let missing_scripts: Vec<ScriptHash> = required_scripts
+        .difference(&provided_scripts)
+        .cloned()
+        .collect();
+
+    if !missing_scripts.is_empty() {
+        return Err(InvalidScripts::MissingRequiredScripts(missing_scripts));
+    }
+
+    let extra_scripts: Vec<ScriptHash> = provided_scripts
+        .difference(&required_scripts)
+        .cloned()
+        .collect();
+    if !extra_scripts.is_empty() {
+        return Err(InvalidScripts::ExtraneousScriptWitnesses(extra_scripts));
+    }
+
+    Ok(())
+}
+
+#[cfg(test)]
+mod tests {
+    use std::ops::Deref;
+
+    use crate::{context::assert::AssertValidationContext, rules::tests::fixture_context};
+    use amaru_kernel::{include_cbor, include_json, MintedTransactionBody, MintedWitnessSet};
+    use test_case::test_case;
+
+    use super::InvalidScripts;
+
+    macro_rules! fixture {
+        ($hash:literal) => {
+            (
+                fixture_context!($hash),
+                include_cbor!(concat!("transactions/preprod/", $hash, "/tx.cbor")),
+                include_cbor!(concat!("transactions/preprod/", $hash, "/witness.cbor")),
+            )
+        };
+        ($hash:literal, $variant:literal) => {
+            (
+                fixture_context!($hash, $variant),
+                include_cbor!(concat!(
+                    "transactions/preprod/",
+                    $hash,
+                    "/",
+                    $variant,
+                    "/tx.cbor"
+                )),
+                include_cbor!(concat!(
+                    "transactions/preprod/",
+                    $hash,
+                    "/",
+                    $variant,
+                    "/witness.cbor"
+                )),
+            )
+        };
+    }
+
+    #[test_case(fixture!("3b54f084af170b30565b1befe25860214a690a6c7a310e2902504dbc609c318e"); "happy path")]
```

Comment:
How are all the json files generated? Would be nice to have some details/scripts shared.

Reply:
Also note that longer term, we'll flesh out the hand-rolled cases by [generating](https://github.com/SundaeSwap-finance/cardano-ledger-conformance-tests/tree/rrruko/dump-test-vectors) golden test vectors and saving them to [Cardano Blueprint](https://github.com/cardano-scaling/cardano-blueprint/pull/41)

The hand-rolled test cases here serve, effectively, as documentation that helps us think through the corner cases, and then then golden test vectors will help us ensure we have high coverage overall.